terror/terror.go

```diff
@@ -0,0 +1,291 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package terror
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+)
+
+// ErrCode represents a specific error type in a error class.
+// Same error code can be used in different error classes.
+type ErrCode int
+
+// ErrCodeText is a textual error code that represents a specific error type in a error class.
+// This string should include component name, see
+// https://github.com/pingcap/tidb/blob/master/docs/design/2020-05-08-standardize-error-codes-and-messages.md#the-error-code-range
+type ErrCodeText string
+
+// ErrClass represents a class of errors.
+type ErrClass int
+
+var errClass2Desc = make(map[ErrClass]string)
+
+// RegisterErrorClass registers new error class for terror.
+func RegisterErrorClass(classCode int, desc string) ErrClass {
```

Comment:
I don't think it's a good idea to make a global registry. Different projects can define their own registry.

Reply:
But projects cannot hook at the `New` or `NewError` function. And if we want to make a generic error document generator, control of registry would be necessary.